pkg/kv/kvserver/store_send.go

```diff
@@ -157,104 +158,166 @@ func (s *Store) Send(
+			startKey := ri.Desc.StartKey
+			if rSpan.Key.Less(startKey) {
+				startKey = rSpan.Key
+			}
+			endKey := ri.Desc.EndKey
+			if endKey.Less(rSpan.EndKey) {
+				endKey = rSpan.EndKey
+			}
+			var ris []roachpb.RangeInfo
+			if err := s.visitReplicasByKey(ctx, startKey, endKey, AscendingKeyOrder, func(ctx context.Context, repl *Replica) error {
+				// Note that we return the lease even if it's expired. The kvclient can use it as it sees fit.
+				ri := repl.GetRangeInfo(ctx)
+				if ri.Desc.RangeID == skipRID {
+					return nil
+				}
+				ris = append(ris, ri)
 				return nil
+			}); err != nil {
+				// Errors here should not be possible, but if there is one, it is ignored
+				// as attaching RangeInfo is optional.
+				log.Warningf(ctx, "unexpected error visiting replicas: %s", err)
+				ris = nil // just to be safe
 			}
-			ris = append(ris, ri)
-			return nil
-		}); err != nil {
-			// Errors here should not be possible, but if there is one, it is ignored
-			// as attaching RangeInfo is optional.
-			log.Warningf(ctx, "unexpected error visiting replicas: %s", err)
-			ris = nil // just to be safe
-		}
-		for _, ri := range ris {
-			t.AppendRangeInfo(ctx, ri)
+			isRetriableRangeKeyMismatch := false
+			candidateRanges = append(candidateRanges, t.Ranges...)
+			for _, ri := range ris {
+				// Update the candidate ranges, that will be combined with the replica batch response RangeInfos
+				// if returned from the replica. Note here that newer ranges are always appended, so that the
+				// oldest RangeInfo is processed by the Client's RangeCache first, which is then invalidated on
+				// conflict by newer data (LILO).
+				ri.ValidateRangeInfoLease(ctx)
+				candidateRanges = append(candidateRanges, ri)
+
+				// Check if the original batch request rSpan exists entirely within any known
+				// ranges stored on this replica, if so update repl and set retriable to true.
+				if ri.Desc.RSpan().ContainsKeyRange(rSpan.Key, rSpan.EndKey) {
+					// Retry this request: Update BatchRequest to reflect re-routing the request
+					// to ri. ClientRangeInfo is also updated here to avoid the replica providing
+					// duplicate BatchResponse.RangeInfos upon a successful retry for the client
+					// to invalidate their RangeCache with.
+					isRetriableRangeKeyMismatch = true
+
+					// Refresh the request and populate the updated request information.
+					baRetry = ba
+					baRetry.RangeID = ri.Desc.RangeID
+					baRetry.ClientRangeInfo.ClosedTimestampPolicy = ri.ClosedTimestampPolicy
+					baRetry.ClientRangeInfo.DescriptorGeneration = ri.Desc.Generation
+					baRetry.ClientRangeInfo.LeaseSequence = ri.Lease.Sequence
+				}
+			}
+
+			// if we are able to retry, iterate retry loop.
+			if isRetriableRangeKeyMismatch {
+				continue
+			}
+
+			// Update the RangeInfos that will be returned to the client, to invalidate
+			// the client's RangeCache.
+			t.AppendRangeInfo(ctx, candidateRanges...)
```

Comment:
Between this and the line above where we populate candidateRanges using t.Ranges:

```go
candidateRanges = append(candidateRanges, t.Ranges...)
```

Are we including duplicates? Do we instead want to write to t.Ranges directly?